In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
#  YOUR CODE GOES HERE
application_df = application_df.drop(columns=['EIN'])

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [4]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
# Cutoff of > 500 data points for APPLICATION_TYPE
application_types_to_replace = ['T9', 'T13', 'T12', 'T2', 'T14', 'T25', 'T15', 'T29', 'T17']

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [5]:
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
classification_counts = application_df['CLASSIFICATION'].value_counts()

# Create a list of classifications to replace (counts < 100)
classifications_to_replace = classification_counts[classification_counts < 100].index

# Replace classifications with "Other" where count < 100
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls, 'Other')

# Check the updated value counts
print(application_df['CLASSIFICATION'].value_counts())


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: count, dtype: int64


In [7]:
# Display value counts for CLASSIFICATION 'Other'
other_classifications_counts = application_df[application_df['CLASSIFICATION'] == 'Other']['CLASSIFICATION'].value_counts()
print(other_classifications_counts)


CLASSIFICATION
Other    669
Name: count, dtype: int64


In [8]:
# Evaluate the NAME Field counts for binning
name_counts = application_df['NAME'].value_counts()
name_counts

NAME
PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: count, Length: 19568, dtype: int64

In [10]:
# Choose a cutoff value and create a list of names to be replaced
names_to_replace = list(name_counts[name_counts < 100].index)

# Replace in dataframe
for name in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(
        name, "Other")

# Check to make sure binning was successful
application_df['NAME'].value_counts()

NAME
Other                                                                 25987
PARENT BOOSTER USA INC                                                 1260
TOPS CLUB INC                                                           765
UNITED STATES BOWLING CONGRESS INC                                      700
WASHINGTON STATE UNIVERSITY                                             492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                         408
PTA TEXAS CONGRESS                                                      368
SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC                           331
ALPHA PHI SIGMA                                                         313
TOASTMASTERS INTERNATIONAL                                              293
MOST WORSHIPFUL STRINGER FREE AND ACCEPTED MASONS                       287
LITTLE LEAGUE BASEBALL INC                                              277
INTERNATIONAL ASSOCIATION OF LIONS CLUBS                                266
MOMS CLUB                                                               210
INTERNATIONAL ASSOCIATION OF SHEET METAL AIR RAIL & TRANSPORTATION      206
AMERICAN ASSOCIATION OF UNIVERSITY WOMEN                                197
FARMERS EDUCATIONAL AND COOPERATIVE UNION OF AMERICA                    166
KNIGHTS OF COLUMBUS                                                     158
HABITAT FOR HUMANITY INTERNATIONAL INC                                  154
TENNESSEE ORDER OF THE EASTERN STAR                                     151
VETERANS OF FOREIGN WARS OF THE UNITED STATES AUXILIARY                 144
PTA UTAH CONGRESS                                                       140
THE UNITED STATES PONY CLUBS INC                                        136
CIVITAN INTERNATIONAL                                                   131
SIGMA BETA DELTA INC                                                    127
HONOR SOCIETY OF PHI KAPPA PHI                                          107
MONTANA 4-H FOUNDATION INC                                              107
WASHINGTON STATE GRANGE                                                 106
UNIVERSITY OF WYOMING                                                   105
DEMOLAY INTERNATIONAL                                                   104
SERTOMA INC                                                             103
Name: count, dtype: int64

In [11]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df_encoded = pd.get_dummies(application_df)

In [12]:
application_df_encoded.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_ALPHA PHI SIGMA  \
0       1     5000              1                 False   
1       1   108590              1                 False   
2       1     5000              0                 False   
3       1     6692              1                 False   
4       1   142590              1                 False   

   NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC  \
0                                              False      
1                                              False      
2                                              False      
3                                              False      
4                                              False      

   NAME_AMERICAN ASSOCIATION OF UNIVERSITY WOMEN  NAME_CIVITAN INTERNATIONAL  \
0                                          False                       False   
1                                          False                       False   
2                                          False                       False   
3                                          False                       False   
4                                          False                       False   

   NAME_DEMOLAY INTERNATIONAL  \
0                       False   
1                       False   
2                       False   
3                       False   
4                       False   

   NAME_FARMERS EDUCATIONAL AND COOPERATIVE UNION OF AMERICA  \
0                                              False           
1                                              False           
2                                              False           
3                                              False           
4                                              False           

   NAME_HABITAT FOR HUMANITY INTERNATIONAL INC  ...  INCOME_AMT_1-9999  \
0                                        False  ...              False   
1                                        False  ...               True   
2                                        False  ...              False   
3                                        False  ...              False   
4                                        False  ...              False   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                   False                     False               False   
1                   False                     False               False   
2                   False                     False               False   
3                    True                     False               False   
4                   False                      True               False   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0             False                   False            False   
1             False                   False            False   
2             False                   False            False   
3             False                   False            False   
4             False                   False            False   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0              False                      True                     False  
1              False                      True                     False  
2              False                      True                     False  
3              False                      True                     False  
4              False                      True                     False  

[5 rows x 81 columns]

In [13]:
# Get the column names from application_df_encoded
print("\nColumns in application_df_encoded:")
for column in application_df_encoded.columns:
    print(column)



Columns in application_df_encoded:
STATUS
ASK_AMT
IS_SUCCESSFUL
NAME_ALPHA PHI SIGMA
NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC
NAME_AMERICAN ASSOCIATION OF UNIVERSITY WOMEN
NAME_CIVITAN INTERNATIONAL
NAME_DEMOLAY INTERNATIONAL
NAME_FARMERS EDUCATIONAL AND COOPERATIVE UNION OF AMERICA
NAME_HABITAT FOR HUMANITY INTERNATIONAL INC
NAME_HONOR SOCIETY OF PHI KAPPA PHI
NAME_INTERNATIONAL ASSOCIATION OF LIONS CLUBS
NAME_INTERNATIONAL ASSOCIATION OF SHEET METAL AIR RAIL & TRANSPORTATION
NAME_KNIGHTS OF COLUMBUS
NAME_LITTLE LEAGUE BASEBALL INC
NAME_MOMS CLUB
NAME_MONTANA 4-H FOUNDATION INC
NAME_MOST WORSHIPFUL STRINGER FREE AND ACCEPTED MASONS
NAME_Other
NAME_PARENT BOOSTER USA INC
NAME_PTA TEXAS CONGRESS
NAME_PTA UTAH CONGRESS
NAME_SERTOMA INC
NAME_SIGMA BETA DELTA INC
NAME_SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC
NAME_TENNESSEE ORDER OF THE EASTERN STAR
NAME_THE UNITED STATES PONY CLUBS INC
NAME_TOASTMASTERS INTERNATIONAL
NAME_TOPS CLUB INC
NAME_UNITED STATES BOWLING CONGRESS 

In [14]:
# Split our preprocessed data into our features and target arrays
y = application_df_encoded['IS_SUCCESSFUL']
X = application_df_encoded.drop(columns=['IS_SUCCESSFUL'])

# Split the preprocessed data into training and testing datasets
# train_test_split from sklearn automatically shuffles and splits the data
# - X contains our feature variables (all columns except IS_SUCCESSFUL)
# - y contains our target variable (IS_SUCCESSFUL column)
# - random_state=42 ensures reproducibility of the split
# - Default test_size=0.25 means 75% training data, 25% testing data
# This split allows us to:
# 1. Train the model on one dataset (X_train, y_train)
# 2. Evaluate performance on separate holdout data (X_test, y_test)
# 3. Avoid overfitting by testing on unseen data
# 4. Random state for reproducibility

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [15]:
# Based on feature importance analysis, dropping less influential columns
# Dropping (12) columns that likely have minimal impact on model accuracy:

# Based on feature importance analysis and data distribution, recommended columns to drop:
# 1. SPECIAL_CONSIDERATIONS_Y - Low variance binary feature
# 2. AFFILIATION_Regional - Less common affiliation type
# 3. CLASSIFICATION_C7000 - Lower frequency classification
# 4. USE_CASE_Other - Generic catch-all category
# 5. INCOME_AMT_50M+ - Very rare income bracket
# 6. APPLICATION_TYPE_T7 - Lower frequency application type
# 7. ORGANIZATION_Association - Organization info captured by AFFILIATION
# 8. STATUS - Binary feature with low predictive power
# 9. CLASSIFICATION_C1700 - Low frequency classification (287 occurrences)
# 10. APPLICATION_TYPE_T8 - Lower frequency application (737 occurrences)
# 11. INCOME_AMT_1-9999 - Can be merged with next bracket
# 12. USE_CASE_Preservation - Less distinctive use case

# Rationale for additional columns:
# - STATUS: Binary features often have limited predictive value
# - CLASSIFICATION_C1700: From distribution analysis, relatively rare class
# - APPLICATION_TYPE_T8: Shows similar patterns to other low-frequency types
# - INCOME_AMT_1-9999: Can be consolidated with higher brackets
# - USE_CASE_Preservation: Shows weaker correlation with target variable


columns_to_drop = ['SPECIAL_CONSIDERATIONS_Y', 'AFFILIATION_Regional',
                  'CLASSIFICATION_C7000', 'USE_CASE_Other',
                  'INCOME_AMT_50M+', 'APPLICATION_TYPE_T7',
                  'ORGANIZATION_Association', 'STATUS',
                  'CLASSIFICATION_C1700', 'APPLICATION_TYPE_T8',
                  'INCOME_AMT_1-9999', 'USE_CASE_Preservation']

# Drop columns from X_train and X_test
X_train_reduced = X_train.drop(columns=columns_to_drop)
X_test_reduced = X_test.drop(columns=columns_to_drop)

# Show the number of columns in both datasets
print("Number of columns in X_train_reduced:", X_train_reduced.shape[1])
print("Number of columns in X_test_reduced:", X_test_reduced.shape[1])

Number of columns in X_train_reduced: 68
Number of columns in X_test_reduced: 68


In [16]:
# Compile, Train, and Evaluate the Model

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=10, activation='relu', input_dim=68))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=8, activation='sigmoid'))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=6, activation='sigmoid'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 10)                  │             690 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 8)                   │              88 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 6)                   │              54 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 1)                   │               7 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 839 (3.28 KB)

 Trainable params: 839 (3.28 KB)

 Non-trainable params: 0 (0.00 B)

In [17]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train_reduced)

# Scale the data
X_train_scaled = X_scaler.transform(X_train_reduced)
X_test_scaled = X_scaler.transform(X_test_reduced)

In [18]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [19]:
# Check the shape of X_train_scaled to see number of features/parameters
print(f"Number of features in X_train_scaled: {X_train_reduced.shape[1]}")
print(f"Number of samples in X_train_scaled: {X_train_reduced.shape[0]}")
print("\nX_train_scaled shape:", X_train_reduced.shape)

Number of features in X_train_scaled: 68
Number of samples in X_train_scaled: 25724

X_train_scaled shape: (25724, 68)


In [20]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.6278 - loss: 0.6587
Epoch 2/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7266 - loss: 0.5469
Epoch 3/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7525 - loss: 0.5250
Epoch 4/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7526 - loss: 0.5190
Epoch 5/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7539 - loss: 0.5167
Epoch 6/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7564 - loss: 0.5075
Epoch 7/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7591 - loss: 0.5037
Epoch 8/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7558 - loss: 0.5015
Epoch 9/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7605 - loss: 0.4993
Epoch 10/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7589 - loss: 0.4955
Epoch 11/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7546 - loss: 0.5009
Epoch 12/100
804/804 ━━━━━━━━━━━━━━━━━━━━

In [21]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - 2ms/step - accuracy: 0.7538 - loss: 0.4942
Loss: 0.49421414732933044, Accuracy: 0.7538192272186279


# Optimized Deep Learning Model Analysis for Charity Success Prediction

## Purpose
This analysis aims to develop a binary classifier using deep learning to help Alphabet Soup predict which funding applicants are most likely to succeed in their ventures. By analyzing historical data from over 34,000 organizations, we seek to create a model that can effectively identify high-potential funding candidates.

## Model Architecture and Implementation

### Data Preprocessing
- Removed non-predictive identification column (EIN)
- Binned rare categories in APPLICATION_TYPE and CLASSIFICATION
- Binned NAME column to reduce noise (binned names with <100 occurrences)
- Standardized features using StandardScaler
- Split data into training and testing sets

### Neural Network Structure
- Input layer: 68 features
- First hidden layer: 10 neurons with ReLU activation
- Second hidden layer: 8 neurons with Sigmoid activation  
- Third hidden layer: 6 neurons with Sigmoid activation
- Output layer: 1 neuron with Sigmoid activation

## Results Analysis

### Model Performance Metrics

- The optimized model showed improved performance compared to the initial model:

 | Metric   | Initial Model | Optimized Model | Improvement |
 |----------|---------------|-----------------|-------------|
 | Loss     | 0.5601       | 0.4942         | -12.2%      |
 | Accuracy | 72.89%       | 75.38%         | +2.61%      |

### Key Questions Addressed

1. **Data Preprocessing**
   - Target variable: IS_SUCCESSFUL
   - Features: All columns except EIN
   - Removed variables: EIN (identification columns)

2. **Compiling, Training, and Evaluating**
   - Neurons: 10, 8, and 6 in three hidden layers
   - Activation functions: ReLU and Sigmoid
   - Target performance: 75% accuracy
   - Steps to increase performance:
     * Added a third hidden layer
     * Adjusted neuron counts
     * Implemented feature reduction and additional binning to reduce noise
     * Enhanced data preprocessing

3. **Model Performance**
   - The optimized model achieved 75.38% accuracy
   - The optimized model met the accuracy goal of 75% 
   - Shows good generalization without overfitting

## Summary

The optimized deep learning model demonstrates reasonable predictive capability with 75.38% accuracy. The optimized model met the 75% target, therefore providing reliable predictions for charity success. The model benefits from careful preprocessing and a balanced architecture that prevents overfitting while maintaining good predictive power.

## Alternative Approach

A Random Forest Classifier could be an effective alternative for this analysis as well:

### Advantages of Random Forest:
- Better handling of categorical variables without extensive preprocessing
- Built-in feature importance rankings
- Less susceptible to outliers
- More interpretable results
- Requires less hyperparameter tuning

The Random Forest approach would be particularly suitable because:
1. The dataset contains many categorical variables
2. We need clear feature importance understanding
3. The binary classification nature of the problem
4. The potential for ensemble methods to capture complex patterns

This alternative could potentially achieve similar or better accuracy while providing more insights into feature importance and decision-making process.
